File: SimulationJava/src/Antylopa.java
import java.util.Random;

import javax.swing.ImageIcon;
import javax.swing.JLabel;

public class Antylopa extends Zwierze {

    static private int probaucieczki=2;

    Antylopa(int xx,int yy, Swiat s){
        x=xx;
        y=yy;
        inicjatywa=4;
        sila=4;
        this.s=s;
        if(s instanceof Hexagony) obrazek = new JLabel(new ImageIcon("antylopahex.png"));
        else obrazek = new JLabel(new ImageIcon("antylopa.png"));
    }

    @Override
    public Organizm dziecko() {
        return new Antylopa(x,y,s);
    }

    @Override
    public void akcja() {
        noweXY nowe = new noweXY(x,y);
        int sx=x,sy=y;
        ruch(nowe);
        if(s.getPole(nowe.nX, nowe.nY)==null){
            tura(nowe);
            int tmpX = nowe.nX;
            int tmpY = nowe.nY;
            nowe.nX = nowe.nX - sx;
		    nowe.nY = nowe.nY - sy;
		    nowe.nX = sx + 2 * nowe.nX;
	    	nowe.nY = sy + 2 * nowe.nY;
		    if (nowe.nX < 0 || nowe.nY < 0 || nowe.nY >= s.getYM() || nowe.nX >= s.getXM()) {
		    	nowe.nX = tmpX;
		        nowe.nY = tmpY;
		    }   
        }
        if(s.getPole(nowe.nX, nowe.nY)!=this && plodnosc==0)tura(nowe);
    }

    @Override
    protected void kolizja(int nX, int nY) {
        makeLog(s.getPole(nX, nY).wypiszNazwe(), "", 3, nX, nY);
        if (s.getPole(nX, nY).getSila() >= sila) {
            Random rand = new Random();
            int ucieczka = rand.nextInt(probaucieczki);
            if (ucieczka == 0) {
                makeLog(s.getPole(nX, nY).wypiszNazwe(), wypiszNazwe(), 2, x, y);
                zycie = 0;
                s.setTab(x, y, s.getPole(nX, nY));
                s.setTab(nX, nY,null);
                s.getPole(x, y).setX(x);
                s.getPole(x, y).setY(y);
            }
            else {
                int Ux = x, Uy = y,g= 0,r;
                int[] t2 = new int[s.getIleS()];
                int[][]t=s.getSasiad(this);

                for(int i=0;i<s.getIleS();i++){
                    if(t[i][0]!=-1){
                        if(s.getPole(t[i][xpos], t[i][ypos])==null){
                            t2[g]=i;
                            g++;
                        }
                    }
                }
                if (g != 0) {
                    r = rand.nextInt(g);
                    Ux = t[t2[r]][xpos];
                    Uy = t[t2[r]][ypos];
                } else {
                    Ux = nX;
                    Uy = nY;
                }
                s.setTab(x,y,s.getPole(nX, nY));
                s.setTab(nX, nY,null);
                s.getPole(x, y).setX(x);
                s.getPole(x, y).setY(y);
                s.setTab(Ux, Uy, this);
                makeLog(wypiszNazwe(), s.getPole(x, y).wypiszNazwe(), 5, Ux, Uy);
                x = Ux;
                y = Uy;
                zycie=mlody;
            }
        }
        else {
            if (s.getPole(nX, nY).getSila() < sila) {
                makeLog(wypiszNazwe(), s.getPole(nX, nY).wypiszNazwe(), 2, x, y);
                s.getPole(nX, nY).setZycie(martwy);
                s.setTab(nX,nY,null);
            }
        }
    }


    @Override
    public String wypiszNazwe() {
        return "antylopa";
    }

    @Override
    public JLabel wypisz() {
        return obrazek;
    }
    
}

File: SimulationCpp/projektPO/antylopa.cpp
#include"antylopa.h"
#include <iostream>
using namespace std;

Antylopa::Antylopa(int x, int y, Swiat* s) {
	sila = 4;
	this->x = x;
	this->y = y;
	inicjatywa = 4;
	zycie = ZYWY;
	this->s = s;
}

void Antylopa::wypisz()const {
	SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), KOLORANTYLOPY);
	cout << "A";
}

void Antylopa::akcja() {
	int nX = x, nY = y, sx = x, sy = y;
	ruch(nX, nY);
	if (s->tab[nX][nY]==nullptr) {
		tura(nX, nY);
		int tmpX = nX;
		int tmpY = nY;
		nX = nX - sx;
		nY = nY - sy;
		nX = sx + 2 * nX;
		nY = sy + 2 * nY;
		if (nX < POCZATEKPLANSZY || nY < POCZATEKPLANSZY || nY >= s->getYM() || nX >= s->getXM()) {
			nX = tmpX;
			nY = tmpY;
		}
	}
	if(s->tab[nX][nY]!=this)tura(nX, nY);
}

void Antylopa::kolizja(int nX, int nY) {
	makeLog(s->tab[nX][nY]->wypiszNazwe(), wypiszNazwe(), s->log, to_string(nX), to_string(nY), KOMz7);
	if (s->tab[nX][nY]->getSila() > sila || s->tab[nX][nY]->getSila() == sila) {
		int ucieczka = rand() % SZANSAUCIECZKI;
		if (ucieczka == NIEUDANA) {
			zycie = MARTWY;
			makeLog(s->tab[nX][nY]->wypiszNazwe(), wypiszNazwe(), s->log, to_string(x), to_string(y), KOMz1);
			s->tab[x][y] = s->tab[nX][nY];
			s->tab[nX][nY] = nullptr;
			s->tab[x][y]->setX(x);
			s->tab[x][y]->setY(y);
		}
		else {
			makeLog(wypiszNazwe(), s->tab[nX][nY]->wypiszNazwe(), s->log, to_string(x), to_string(y), KOMz4);
			int Ux = x, Uy = y;
			bool znaleziono = false;
			for (int i = SASIADGORA; i < SASIADDOL; i++) {
				if (x + i < POCZATEKPLANSZY || x + i >= s->getXM()) continue;
				for (int j = SASIADLEWO; j < SASIADPRAWO; j++) {
					if (y + j < POCZATEKPLANSZY || y + j >= s->getYM()) continue;
					if ((i != SRODEK && j == SRODEK) || (i == SRODEK && j != SRODEK)) {
						if (!s->tab[x + i][y + j]) {
							znaleziono = true;
							Ux = x + i;
							Uy = y + j;
							break;
						}
					}
				}
				if (znaleziono) break;
			}
			if (!znaleziono) {
				Ux = nX;
				Uy = nY;
			}
			s->tab[x][y] = s->tab[nX][nY];
			s->tab[nX][nY] = nullptr;
			s->tab[x][y]->setX(x);
			s->tab[x][y]->setY(y);
			s->tab[Ux][Uy] = this;
			x = Ux;
			y = Uy;
		}
	}
	else {
		if (s->tab[nX][nY]->getSila() < sila) {
			makeLog(wypiszNazwe(), s->tab[nX][nY]->wypiszNazwe(), s->log, to_string(x), to_string(y), KOMz1);
			s->tab[nX][nY]->zycie = MARTWY;
			s->tab[nX][nY] = nullptr;
		}
	}
}

string Antylopa::wypiszNazwe() const {
	return "antylopa";
}
Organizm* Antylopa::dziecko() {
	return new Antylopa(x, y, s);
}









